### PR TITLE
dialects: Switch all arith binary operations to init

### PR DESCRIPTION
**Pull request by PapyChacal**
Switches all arith binary operations to use _init_, because it's factored out now!

### CODE REVIEW

**Review by AntonLydike — APPROVED** [review]
Coolio!

File: xdsl/dialects/arith.py [hunk 1]
```diff
@@ -153,10 +153,15 @@ class BinaryOperation(IRDLOperation, Generic[_T]):
     rhs: Annotated[Operand, _T]
     result: Annotated[OpResult, _T]
 
-    @classmethod
-    def get(cls, operand1: Operation | SSAValue, operand2: Operation | SSAValue):
-        operand1 = SSAValue.get(operand1)
-        return cls.build(operands=[operand1, operand2], result_types=[operand1.typ])
+    def __init__(
+        self,
+        operand1: Union[Operation, SSAValue],
```
Comment (math-fehr) [inline xdsl/dialects/arith.py:158]:
```suggestion
        operand1: Operation | SSAValue,
```

File: xdsl/dialects/arith.py [hunk 2]
```diff
@@ -153,10 +153,15 @@ class BinaryOperation(IRDLOperation, Generic[_T]):
...
+        operand2: Union[Operation, SSAValue],
```
Comment (math-fehr) [inline xdsl/dialects/arith.py:159]:
```suggestion
        operand2: Operation | SSAValue,
```

File: xdsl/dialects/arith.py (hunk 1 above)
Comment (math-fehr) [reply xdsl/dialects/arith.py:158]: I guess I missed that on the `Addi.__init__`

**Review by math-fehr — APPROVED** [review]
Nice!

File: tests/dialects/test_arith.py [hunk 3]
```diff
@@ -49,7 +49,7 @@
 from xdsl.ir import Attribute
 from xdsl.utils.exceptions import VerifyException
 
-_BinOpT = TypeVar("_BinOpT", bound=BinaryOperation)
+_BinOpT = TypeVar("_BinOpT", bound=Attribute)
```
Comment (superlopuh) [inline tests/dialects/test_arith.py:52]: why this change?

**Review by superlopuh — APPROVED** [review]

**Review by webmiche — APPROVED** [review]
Nice, LGTM!
Imagine we had used the test dialect all the testing, the you wouldn't have had to toich as many files :wink:

**Review by georgebisbas — APPROVED** [review]

File: tests/dialects/test_arith.py (hunk 3 above)
Comment (PapyChacal) [reply tests/dialects/test_arith.py:52]: This is basically Pyright shennanigans in Pytest parameterized test. The test is parameterized by the Operation class (Addi, Subi, ...). It used to be happy with `type[_BinOpT]` (I would argue it was overkill and `type[BinaryOperation]` should have yielded the same thing, but I didn't play around)
Now that it's generic, fields and such are actually defined by `BinaryOperation`, so it makes Pyright happier. But it requires a type parameter! So I switch to `type[BinaryOperation[_binOpT]]`
